py/h2o_wave/h2o_wave/cli.py

```diff
@@ -143,40 +144,50 @@ def run(app: str, no_reload: bool, no_autostart: bool):
 
     # Try to start Wave daemon if not running or turned off.
     server_port = int(os.environ.get('H2O_WAVE_LISTEN', ':10101').split(':')[-1])
-    server_not_running = _scan_free_port(server_port) == server_port
+    server_running = _scan_free_port(server_port) != server_port
 
     waved_process = None
     if no_autostart:
         autostart = False
     else:
         autostart = os.environ.get('H2O_WAVE_NO_AUTOSTART', 'false').lower() in ['false', '0', 'f']
 
-    waved_path = os.path.join(sys.exec_prefix, 'waved.exe' if IS_WINDOWS else 'waved')
+    waved_cwd = sys.exec_prefix
+    proc_name = 'waved.exe' if IS_WINDOWS else 'waved'
+    # If waved is running, find its cwd.
+    if server_running:
+        for proc in psutil.process_iter():
+            if proc.name() == proc_name:
+                try:
+                    waved_cwd = psutil.Process(proc.pid).cwd()
+                except psutil.NoSuchProcess:
+                    pass
+    waved_path = os.path.join(waved_cwd, proc_name)
     # OS agnostic wheels do not include waved - needed for HAC.
     is_waved_present = os.path.isfile(waved_path)
 
     try:
-        if autostart and is_waved_present and server_not_running:
+        if autostart and is_waved_present and not server_running:
             kwargs = {}
             if IS_WINDOWS:
                 kwargs['creationflags'] = subprocess.CREATE_NEW_PROCESS_GROUP
 
             waved_process = subprocess.Popen([waved_path], cwd=sys.exec_prefix, env=os.environ.copy(), **kwargs)
             time.sleep(1)
-            server_not_running = _scan_free_port(server_port) == server_port
+            server_running = _scan_free_port(server_port) != server_port
             retries = 3
-            while retries > 0 and server_not_running:
+            while retries > 0 and not server_running:
                 print('Cannot connect to Wave server, retrying...')
                 time.sleep(2)
-                server_not_running = _scan_free_port(server_port) == server_port
+                server_running = _scan_free_port(server_port) != server_port
                 retries = retries - 1
 
-        if autostart and server_not_running:
+        if autostart and not server_running:
             print('Could not connect to Wave server. Please start the Wave server (waved or waved.exe) prior to running any app.')
             return
 
         if not os.environ.get('H2O_WAVE_WAVED_DIR') and is_waved_present:
-            os.environ['H2O_WAVE_WAVED_DIR'] = sys.exec_prefix
+            os.environ['H2O_WAVE_WAVED_DIR'] = waved_cwd
```

Comment:
If this works reliably, we can deprecate `H2O_WAVE_WAVED_DIR`. Add a TODO here for now.